internal/cli/command/bucket/create.go

```diff
@@ -0,0 +1,261 @@
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/antihax/optional"
+	"github.com/goph/emperror"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"gopkg.in/AlecAivazis/survey.v1"
+
+	"github.com/banzaicloud/banzai-cli/.gen/pipeline"
+	"github.com/banzaicloud/banzai-cli/internal/cli"
+	"github.com/banzaicloud/banzai-cli/internal/cli/format"
+	"github.com/banzaicloud/banzai-cli/internal/cli/input"
+	"github.com/banzaicloud/banzai-cli/internal/cli/utils"
+)
+
+type createBucketsOptions struct {
+	name     string
+	cloud    string
+	secretID string
+
+	location string
+
+	storageAccount string
+	resourceGroup  string
+}
+
+// NewCreateCommand creates a new cobra.Command for `banzai bucket create`.
+func NewCreateCommand(banzaiCli cli.Cli) *cobra.Command {
+	o := createBucketsOptions{}
+
+	cmd := &cobra.Command{
+		Use:     "create NAME [[--cloud=]CLOUD] [[--location=]LOCATION] [[--secret-id=]SECRET_ID]",
+		Short:   "Create bucket",
+		Long:    "Create object storage bucket on supported cloud providers",
+		Args:    cobra.MaximumNArgs(4),
+		Aliases: []string{"c", "create"},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cmd.SilenceErrors = true
+			cmd.SilenceUsage = true
+
+			if len(args) > 0 {
+				o.name = args[0]
+			}
+			if len(args) > 1 {
+				o.cloud = args[1]
+			}
+			if len(args) > 2 {
+				o.location = args[2]
+			}
+			if len(args) > 3 {
+				o.secretID = args[3]
+			}
+
+			if !banzaiCli.Interactive() {
+				if o.name == "" {
+					return errors.New("NAME must be specified")
+				}
+				if o.cloud == "" {
+					return errors.New("CLOUD argument or --cloud flag must be specified")
+				}
+				if o.location == "" {
+					return errors.New("LOCATION argument or --location flag must be specified")
+				}
+				if o.secretID == "" {
+					return errors.New("SECRET_ID argument or --secret-id flag must be specified")
+				}
+			}
+
+			if o.cloud == input.CloudProviderAzure {
+				if o.storageAccount == "" {
+					return errors.New("--storage-account must be specified for Azure")
+				}
+				if o.resourceGroup == "" {
+					return errors.New("--resource-group must be specified for Azure")
+				}
+			}
+
+			return runCreate(banzaiCli, o)
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.StringVarP(&o.cloud, "cloud", "c", "", "Cloud provider for the bucket")
+	flags.StringVarP(&o.location, "location", "l", "", "Location for the bucket")
+	flags.StringVarP(&o.secretID, "secret-id", "s", "", "Secret ID of the used secret to create the bucket")
+	flags.StringVarP(&o.storageAccount, "storage-account", "", "", "Storage account for the bucket (must be specified for Azure)")
+	flags.StringVarP(&o.resourceGroup, "resource-group", "", "", "Resource group for the bucket (must be specified for Azure)")
+
+	return cmd
+}
+
+func runCreate(banzaiCli cli.Cli, o createBucketsOptions) error {
+	var err error
+
+	defaultBucketName := fmt.Sprintf("bucket-%s", strconv.FormatInt(time.Now().UTC().Unix(), 10))
```

Comment:
might use $USER instead of "bucket"?

Reply:
If we use $USER than we have to ensure that the value of it a valid bucket name on each provider.

Reply:
We can't know it before the bucket is actually created (at least because of the possible conflicts). UNIX usernames are of a quite conservative set of chars, so we can be pretty sure though. This is only a default, and the user will get a correct error message if their name is too special for a provider. The same applies to cluster names, where accepting the default/generated name is far from production/intended use of the tool, but will probably work.

Reply:
I think it is safe to go with $USER, the bucket name is pre-validated (although we most probably need to double test the regexes).